platforms/ide/problems-api/src/main/java/org/gradle/api/problems/AdditionalData.java

```diff
@@ -16,26 +16,28 @@
 
 package org.gradle.api.problems;
 
-import org.gradle.api.Action;
 import org.gradle.api.Incubating;
 import org.gradle.api.problems.internal.InternalProblemSpec;
 
+import java.io.Serializable;
+
 /**
  * Marker interface for additional data that can be attached to a {@link Problem}.
+ * It and all its contained Objects need to be Serializable to be able to be carried in a BuildEvent to the TAPI.
+ * We use the {@link org.gradle.tooling.internal.provider.serialization.PayloadSerializer} to serialize the event.
+ * If the serialization fails the build TAPI call will fail.
```

Comment:
Let's not make `PayloadSerializer` part of the public API. I'll try to be explicit what should  be here:
- Concrete classses need to be `Serializable` as the data potentially need to be transferred accross JVMs (for example in case of tooling). 
- _some_ java Serialization based approach is used for serialization (should not be specified the logic exactly)
- Serialization is done in a best effort manner. If additional data serialization fails, problems will be still sent without.